oauth2_client/client.py

```diff
@@ -43,63 +101,68 @@ def request(self, method, url, *args, **kwargs):  # pylint: disable=arguments-di
+                raise
+            new_token = self.expired_callback(self.app)
+            self.token = new_token.to_client_dict()
+            log.debug("Obtained new token. Retrying request.")
+            return self.request(method, url, *args, **kwargs)
+
+
+def get_client(app_name):
     """
-    Returns client to make HTTP requests with OAuth2 access token to the service with the given service_id.
+    Returns HTTP(S) client for authenticated communication with Resource Owner specified by the
+    `app_name` parameter. Identification is by means of OAuth token.
 
-    The access token is loaded from the database if there is an active (not expired) one. It's fetched from the
-    provider service by HTTP and stored in the database otherwise.
+    The access token is loaded from the database, if there is a valid one.
+    Otherwise - new token is fetched from the auth provider by HTTP(S) and stored in the database.
+    The new token is then used for communication. Tokens are automatically refreshed by repeating
+    the authorization flow.
+
+    The client may occasionally return HTTP 401 when token expires during request processing on
+    Resource Owner side.
 
     Arguments:
-        service_name (str): name of the service make requests to f.e. licence.
+        app_name (str): name of the OAuth client application to make requests to f.e. license.
 
     Returns:
-        Ouath2 client to make HTTP requests.
-        The client request could return HTTP 401 for expired token if the request takes too long, and the token
-        is expired till the token check.
-    """
-    access_token = AccessToken.objects.filter(application__name=service_name).order_by('-created_at').first()
-    if access_token is None or _expired(access_token.token):
-        if access_token is not None:
-            app = access_token.application
-        else:
-            app = Application.objects.get(name=service_name)
-        token = _get_token(app)
-        AccessToken.objects.create(application=app, token=token)
-    else:
-        app = access_token.application
-        token = access_token.token
-
-    return OAuth2Client(app, token)
-
-
-def _expired(access_token):
+        client (oauth2_client.OAuth2Client): OAuth2 client for authenticated HTTP(S) communication
     """
-    Indicates if the given access token is almost expired (left less than TIMEOUT_SECONDS till expiration time).
+    token = AccessToken.objects.filter(application__name=app_name).order_by('-created').first()
+    if not token or token.is_expired():
+        app = token.application if token else Application.objects.get(name=app_name)
+        token = fetch_and_store_token(app)
 
-    Arguments:
-        access_token (object): oauth2_client.models.AccessToken instance.
-    """
-    return time.time() > access_token['expires_at'] - TIMEOUT_SECONDS
+    return OAuth2Client(token, expired_callback=fetch_and_store_token)
 
 
-def _get_token(app):
+def fetch_and_store_token(app):
```

Comment:
I do not need we need this as separate function, but maybe you need it in tests?

Reply:
This is widely used as a separate function, for mocking in tests and now as a protection point with a circuit breaker.